### ISSUE TITLE
Boto 2.42 `get_all_instances` API returns null result in case there are more than 10 tags specified in the `filters`

### ISSUE THREAD

**Issue by romanrev** [boto/boto#3637]
Boto version: boto (2.42.0), botocore (1.4.58)
Python: 2.7.12

We have recently discovered that in case we use more than 10 tag/value pairs in the call to get_all_instances from boto.ec2_connection.EC2Connection we get a null result instead of the list of instances. Below is an example:

	>>> from aws_common import get_ec2_connection
	>>> ec2_conn = get_ec2_connection('ap-southeast-2')
	>>> tags = [ { 'Division': ''}, {'Name': 'SYDAWSFSS02'}, {'Application': 'Vedaleon_APIGateway'}, {'CostCentre': 'Ground Ops - FSS'}, {'Environment': 'Prod'}, {'ManagedBy': ''}, {'CreatedBy': ''}, {'Owner': ''}, {'Backup': '1d+5d+all'}, {'Availability': '24x7'}, {'Server Name': 'FSS Server B'} ]
	>>> filters = reduce(lambda r, d: r.update({"tag:"+d.keys()[0]: d.values()[0]}) or r, tags, {})
	>>> filters
	{'tag:Backup': '1d+5d+all', 'tag:Server Name': 'FSS Server B', 'tag:CostCentre': 'Ground Ops - FSS', 'tag:ManagedBy': '', 'tag:Owner': '', 'tag:Application': 'Vedaleon_APIGateway', 'tag:Environment': 'Prod', 'tag:Name': 'SYDAWSFSS02', 'tag:Availability': '24x7', 'tag:CreatedBy': '', 'tag:Division': ''}
	>>> reservations = ec2_conn.get_all_instances(filters=filters)
	>>> instances = [i for r in reservations for i in r.instances]
	>>> instances
	[]

As you can see, in case there are 11 tags we get an empty list of instances. Now if we remove an arbitrary tag value pair, say Availability, reducing the number of tag/value pairs in the filter to 10, we get the instance we are after:

	>>> tags = [ { 'Division': ''}, {'Name': 'SYDAWSFSS02'}, {'Application': 'Vedaleon_APIGateway'}, {'CostCentre': 'Ground Ops - FSS'}, {'Environment': 'Prod'}, {'ManagedBy': ''}, {'CreatedBy': ''}, {'Owner': ''}, {'Backup': '1d+5d+all'}, {'Server Name': 'FSS Server B'} ]
	>>> filters = reduce(lambda r, d: r.update({"tag:"+d.keys()[0]: d.values()[0]}) or r, tags, {})
	>>> reservations = ec2_conn.get_all_instances(filters=filters)
	>>> instances = [i for r in reservations for i in r.instances]
	>>> instances
	[Instance:i-df5e2003]

Below is the reference to the `get_ec2_connection` and `get_aws_access` functions we are using in the example:

	def get_ec2_connection(region=None, role_arn=None, profile=None):
	    """ Obtain an EC2 connection: direct or by assuming a role
	    """
	    aws_access=get_aws_access(region, role_arn, profile)
	    ec2_conn = boto.ec2.connect_to_region(aws_access['region'],
	            aws_access_key_id=aws_access['key'],
	            aws_secret_access_key=aws_access['secret'],
	            security_token=aws_access['token'],
	            profile_name=aws_access['profile']
	            )
	    return ec2_conn

	def get_aws_access(region=None, role_arn=None, profile=None):
	    """ Populate dict for AWS connection
	    """
	    aws_access = {}
	    if region is None:
	        instance_metadata = boto.utils.get_instance_identity()
	        region = instance_metadata['document']['region']
	    if role_arn is None:
	        aws_access['key']=None
	        aws_access['secret']=None
	        aws_access['token']=None
	    else:
	        sts_connection = boto.sts.STSConnection()
	        role_obj = sts_connection.assume_role(
	            role_arn=role_arn,
	            role_session_name="AssumeRoleSession"
	            )
	        aws_access['key']=role_obj.credentials.access_key
	        aws_access['secret']=role_obj.credentials.secret_key
	        aws_access['token']=role_obj.credentials.session_token
	    aws_access['profile'] = profile
	    aws_access['region']=region
	    return aws_access

**Comment by stealthycoin** [comment]
I got the same behavior with boto3, informed the ec2 service team.

**Comment by romanrev** [comment]
Yes @stealthycoin I opened another issue against the Boto 3 repo as well: https://github.com/boto/boto3/issues/880
I don't understand though why you have added the label "closing soon if no response".
